miniapps/multidomain/multidomain.cpp

```diff
@@ -0,0 +1,374 @@
+      {
+         M.Reset(Mform.ParallelAssemble(), true);
+
+         inflow = new ConstantCoefficient(0.0);
+         bform.AddBdrFaceIntegrator(
+            new BoundaryFlowIntegrator(*inflow, *q, alpha));
+      }
+      else
+      {
+         Kform.AddDomainIntegrator(new ConvectionIntegrator(*q, -alpha));
+         Kform.AddDomainIntegrator(new DiffusionIntegrator(*d));
+         Kform.Assemble(0);
+
+         Array<int> empty;
+         Kform.FormSystemMatrix(empty, K);
+         Mform.FormSystemMatrix(ess_tdofs_, M);
+
+         bform.Assemble();
+         b = bform.ParallelAssemble();
+      }
+
+      M_solver.iterative_mode = false;
+      M_solver.SetRelTol(1e-8);
+      M_solver.SetAbsTol(0.0);
+      M_solver.SetMaxIter(100);
+      M_solver.SetPrintLevel(0);
+      M_prec.SetType(HypreSmoother::Jacobi);
+      M_solver.SetPreconditioner(M_prec);
+      M_solver.SetOperator(*M);
+
+      t1.SetSize(height);
+      t2.SetSize(height);
+   }
+
+   void Mult(const Vector &u, Vector &du_dt) const override
+   {
+      K->Mult(u, t1);
+      t1.Add(1.0, *b);
+      M_solver.Mult(t1, du_dt);
+      du_dt.SetSubVector(ess_tdofs_, 0.0);
+   }
+
+   ~ConvectionDiffusionTDO()
+   {
+      delete q;
+      delete d;
+      delete b;
+   }
+
+   /// Mass form
+   ParBilinearForm Mform;
+
+   /// Stiffness form. Might include diffusion, convection or both.
+   ParBilinearForm Kform;
+
+   /// Mass opeperator
+   OperatorHandle M;
+
+   /// Stiffness opeperator. Might include diffusion, convection or both.
+   OperatorHandle K;
+
+   /// RHS form
+   ParLinearForm bform;
+
+   /// RHS vector
+   Vector *b = nullptr;
+
+   /// Velocity coefficient
+   VectorCoefficient *q = nullptr;
+
+   /// Diffusion coefficient
+   Coefficient *d = nullptr;
+
+   /// Inflow coefficient
+   Coefficient *inflow = nullptr;
+
+   /// Essential true dof array. Relevant for eliminating boundary conditions
+   /// when using an H1 space.
+   Array<int> ess_tdofs_;
+
+   double current_dt = -1.0;
+
+   /// Mass matrix solver
+   CGSolver M_solver;
+
+   /// Mass matrix preconditioner
+   HypreSmoother M_prec;
+
+   /// Auxiliary vectors
+   mutable Vector t1, t2;
+};
+
+int main(int argc, char *argv[])
+{
+   Mpi::Init();
+   Hypre::Init();
+
+   int num_procs = Mpi::WorldSize();
+   int myid = Mpi::WorldRank();
+
+   Mesh *serial_mesh = new Mesh("data/multidomain-hex.mesh");
+   ParMesh parent_mesh = ParMesh(MPI_COMM_WORLD, *serial_mesh);
+   delete serial_mesh;
+
+   parent_mesh.UniformRefinement();
+
+   int p = 2;
+   H1_FECollection fec(p, parent_mesh.Dimension());
+
+   // Create the sub-domains and accompanying Finite Element spaces from
+   // corresponding attributes. This specific mesh has two domain attributes and
+   // 9 boundary attributes.
+   Array<int> cylinder_domain_attributes(1);
+   cylinder_domain_attributes[0] = 1;
+
+   auto cylinder_submesh =
+      ParSubMesh::CreateFromDomain(parent_mesh, cylinder_domain_attributes);
```

Comment:
Just for my understanding: If I have really large problem with different interacting physics, then the idea is to pre-partition the mesh (in a balanced way) and replace the code above with loading parallel portion of the mesh from the file, figure out which subdomains are visible on the local partition and then I simply call `CreateFromDomain` on the locally visible integrators/forms?

Reply:
`ParSubMesh::CreateFromDomain` figures that out for you. What it doesn't do for you is "optimal" partitioning of the created `ParSubMesh`.